src/ApiCore/GapicClientTrait.php

```diff
@@ -95,117 +99,193 @@ private static function getGapicVersion(array $options)
             'libVersion' => $this->pluck('libVersion', $options, false),
-            'gapicVersion' => $options['gapicVersion'],
+            'gapicVersion' => $gapicVersion,
         ]);
 
+        self::validateFileExists($options['descriptorsConfigPath']);
         $descriptors = require($options['descriptorsConfigPath']);
         $this->descriptors = $descriptors['interfaces'][$this->serviceName];
 
-        if (isset($options['credentialsLoader'])) {
-            $authHttpHandler = isset($options['authHttpHandler']) ? $options['authHttpHandler'] : null;
-            $this->authWrapper = new AuthWrapper($options['credentialsLoader'], $authHttpHandler);
-        } else {
-            $this->authWrapper = AuthWrapper::build($options);
-        }
+        $serviceAddress = $options['serviceAddress'];
+        $auth = $options['auth'];
+        $authConfig = $options['authConfig'] ?: [];
+        $this->authWrapper = $this->createAuthWrapper($auth, $authConfig);
 
         $this->transport = $transport instanceof TransportInterface
             ? $transport
-            : TransportFactory::build($options);
+            : $this->createTransport($serviceAddress, $transport, $transportConfig);
+    }
+
+    /**
+     * @param mixed $auth
+     * @param array $authConfig
+     * @return AuthWrapper
+     * @throws ValidationException
+     */
+    private function createAuthWrapper($auth, array $authConfig)
+    {
+        if (is_null($auth)) {
+            return AuthWrapper::build($authConfig);
+        } elseif (is_string($auth) || is_array($auth)) {
+            return AuthWrapper::fromKeyFile($auth, $authConfig);
+        } elseif ($auth instanceof FetchAuthTokenInterface) {
+            $authHttpHandler = isset($authConfig['authHttpHandler'])
+                ? $authConfig['authHttpHandler']
+                : null;
+            return new AuthWrapper($auth, $authHttpHandler);
+        } elseif ($auth instanceof AuthWrapper) {
+            return $auth;
+        } else {
+            throw new ValidationException(
+                'Unexpected value in $auth option, got: ' .
+                print_r($auth, true)
+            );
+        }
+    }
+
+    /**
+     * @param string $serviceAddress
+     * @param string $transport
+     * @param array $transportConfig
+     * @return TransportInterface
+     * @throws ValidationException
+     */
+    private function createTransport($serviceAddress, $transport, array $transportConfig)
+    {
+        if (!is_string($transport)) {
+            throw new ValidationException(
+                "'transport' must be a string, instead got:" .
+                print_r($transport, true)
+            );
+        }
+        $configForSpecifiedTransport = isset($transportConfig[$transport])
+            ? $transportConfig[$transport]
+            : [];
+        switch ($transport) {
+            case 'grpc':
+                self::validateGrpcSupport();
```

Comment:
This already happens in [`GrpcTransport::build()`](https://github.com/michaelbausor/gax-php/blob/d7cdbdfaedaa4bd74b2306e9322a35b8b8d17e2c/src/ApiCore/Transport/GrpcTransport.php#L73)